front/lib/resources/user_resource.ts

```diff
@@ -0,0 +1,147 @@
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class UserResource extends BaseResource<User> {
+  static model: ModelStatic<User> = User;
+
+  constructor(model: ModelStatic<User>, blob: Attributes<User>) {
+    super(User, blob);
+  }
+
+  static async findAllByUsername(username: string): Promise<UserResource[]> {
+    const users = await User.findAll({
+      where: {
+        username,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchByExternalId(
+    sid: string,
+    transaction?: Transaction
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        sId: sid,
+      },
+      transaction,
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByAuth0Sub(
+    sub: string,
+    transaction?: Transaction
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        auth0Sub: sub,
+      },
+      transaction,
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByEmail(
+    email: string,
+    transaction?: Transaction
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        email,
+      },
+      transaction,
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByProvider(
+    provider: UserProviderType,
+    providerId: string,
+    transaction?: Transaction
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        provider,
+        providerId,
+      },
+      transaction
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async getWorkspaceFirstAdmin(
+    workspaceId: number
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      include: [
+        {
+          model: MembershipModel,
+          where: {
+            role: "admin",
+            workspaceId,
+          },
+          required: true,
+        },
+      ],
+      order: [["createdAt", "ASC"]],
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  async update(blob: Partial<Attributes<User>>): Promise<void> {
+    const [, affectedRows] = await this.model.update(blob, {
+      where: {
+        id: this.id,
+      },
+      returning: true,
+    });
+
+    // Update the resource to reflect the new status.
+    Object.assign(this, affectedRows[0].get());
+  }
+
+  async delete(
+    auth: Authenticator,
+    transaction?: Transaction
+  ): Promise<Result<undefined, Error>> {
+    try {
+      await this.model.destroy({
+        where: {
+          id: this.id,
+        },
+        transaction,
+      });
+
+      return new Ok(undefined);
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  toJSON(): User {
```

Comment:
Nope toJSON should return a `UserType` not `User` which is not a serializable thing